[AArch64][PAC] Select auth+load into LDRAA/LDRAB/LDRA[pre].

ahmedbougacha opened the pull request:
This can lower loads of a ptrauth.auth base into a fixed sequence that doesn't allow the raw intermediate value to be exposed.

It's based on the AArch64 LDRAA/LDRAB instructions, but as those have limited encodings (in particular, small immediate offsets, and only zero discriminators), it generalizes them with a LDRA pseudo.

It handles arbitrary ptrauth schemas on the authentication, materializing the integer constant discriminator and blending it with an address discriminator if needed.

It handles arbitrary offsets (applied after the authentication).

It also handles pre-indexing with writeback, either writing back the authentication result alone if the offset is 0, or both authentication and offset addition otherwise.

At ISel time, the real LDRAA family of instructions is selected when possible, to avoid needlessly constraining regalloc with X16/X17. After ISel, the LDRA pseudos are expanded in AsmPrinter, into either of:
- writeback, 0 offset (we already wrote the AUT result): LDRXui
- no wb, uimm12s8 offset (including 0): LDRXui
- no wb, simm9 offset: LDURXi
- pre-indexed wb, simm9 offset: LDRXpre
- no wb, any offset: expanded MOVImm + LDRXroX
- pre-indexed wb, any offset: expanded MOVImm + ADD + LDRXui

Though the main intended optimization target is vtable-like codegen, where both the base vtable pointer is signed, as well as its entries, at small fixed offsets.  This does benefit from writeback, hence the ISel complexity to support that, as it's otherwise unlikely to be worthwhile.

GlobalISel would benefit from further optimization, as this lowering conflicts with the generic indexed lowering there.

I did a pass to refresh the old patch, but it's been a while, please let me know if I missed a spot!

ahmedbougacha committed:
[AArch64][PAC] Select auth+load into LDRAA/LDRAB/LDRA[pre].

This can lower loads of a ptrauth.auth base into a fixed sequence that
doesn't allow the raw intermediate value to be exposed.

It's based on the AArch64 LDRAA/LDRAB instructions, but as those have
limited encodings (in particular, small immediate offsets, and only zero
discriminators), it generalizes them with a LDRA pseudo.

It handles arbitrary ptrauth schemas on the authentication,
materializing the integer constant discriminator and blending it with an
address discriminator if needed.

It handles arbitrary offsets (applied after the authentication).

It also handles pre-indexing with writeback, either writing back the
authentication result alone if the offset is 0, or both authentication
and offset addition otherwise.

At ISel time, the real LDRAA family of instructions is selected when
possible, to avoid needlessly constraining regalloc with X16/X17.
After ISel, the LDRA pseudos are expanded in AsmPrinter, into either of:
- writeback, 0 offset (we already wrote the AUT result): LDRXui
- no wb, uimm12s8 offset (including 0): LDRXui
- no wb, simm9 offset: LDURXi
- pre-indexed wb, simm9 offset: LDRXpre
- no wb, any offset: expanded MOVImm + LDRXroX
- pre-indexed wb, any offset: expanded MOVImm + ADD + LDRXui

Though the main intended optimization target is vtable-like codegen,
where both the base vtable pointer is signed, as well as its entries, at
small fixed offsets.  This does benefit from writeback, hence the ISel
complexity to support that, as it's otherwise unlikely to be worthwhile.

GlobalISel would benefit from further optimization, as this lowering
conflicts with the generic indexed lowering there.

atrosinenko commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -1671,6 +1673,163 @@ bool AArch64DAGToDAGISel::tryIndexedLoad(SDNode *N) {
+    SDValue Key = CurDAG->getTargetConstant(KeyC, DL, MVT::i32);
+    SDValue Ops[] = {Offset, Key, IntDisc, AddrDisc, X16Copy.getValue(1)};
+    Res = CurDAG->getMachineNode(Opc, DL, MVT::i64, MVT::Other, MVT::Glue, Ops);
+    if (needsWriteback)
+      Writeback = CurDAG->getCopyFromReg(SDValue(Res, 1), DL, AArch64::X16,
+                                         MVT::i64, SDValue(Res, 2));
+    ResVal = SDValue(Res, 0);
+    OutChain = SDValue(Res, 1);
```
Do I understand correctly: strictly speaking, `OutChain` should be `SDValue(Writeback, 1)` if `needsWriteback` is set, but `Writeback` is scheduled "immediately after LDRApre" and all other successors are simply "after LDRApre", so everything is scheduled correctly without further complicating the computations of `OutChain`?

ahmedbougacha replied:
Yeah, mostly, with two small complications: `Writeback` only has data dependent users (so doesn't need to express control dependents with the chain);  and the control dependency from the `Writeback` `CopyFromReg` to the load is expressed with the somewhat-unusual "Glue" SDValue (here `SDValue(Res, 2)`, per the `getMachineNode` having a `MVT::Glue` result type, in addition to the `MVT::Other` used for the chain.
(I haven't touched SDAG much in a long time, and this is from admittedly old recollections and re-reading the code now, so I may be forgetting something.  I would expect these glue shenanigans to be relatively idiomatic for anything that does copy to/from reg in SDAG, to avoid interference with some unrelated subgraph of nodes that also writes to the same register but has no common control dependencies)

atrosinenko commented on llvm/lib/Target/AArch64/AArch64InstrGISel.td:
```diff
@@ -25,6 +25,23 @@ def G_ADD_LOW : AArch64GenericInstruction {
+def G_LDRA : AArch64GenericInstruction {
+  let OutOperandList = (outs type0:$dst);
+  let InOperandList = (ins type1:$addr, i64imm:$offset, i32imm:$key, i64imm:$disc, type0:$addrdisc);
+  let hasSideEffects = 0;
```
Spotted that in GISel, `G_LDRA` and `G_LDRApre` both define `hasSideEffects` to false. In DAGISel, on the other hand, both `LDRA` and `LDRApre` define `hasSideEffects` to true. Is that intended?

ahmedbougacha committed:
Merge branch 'main' into users/ahmedbougacha/ptrauth-load-auth-ldra

ahmedbougacha committed:
Address review nits.

ahmedbougacha committed:
Merge branch 'main' into users/ahmedbougacha/ptrauth-load-auth-ldra

ahmedbougacha committed:
Update call test for updated auth codegen.

The test didn't need an auth, and could use an x16-defining
inlineasm or something, but the bit it's testing didn't change.

ahmedbougacha committed:
Address review comments, test ELF, and simplify a bit.
